ebos/collecttoiorank.cc

```diff
@@ -0,0 +1,902 @@
+    PackUnPackCellData packUnpackCellData {
+        localCellData,
+        this->globalCellData_,
+        this->localIndexMap_,
+        this->indexMaps_,
+        this->numCells(),
+        this->isIORank()
+    };
+
+    if (! isParallel()) {
+        // no need to collect anything.
+        return;
+    }
+
+    PackUnPackWellData packUnpackWellData {
+        localWellData,
+                this->globalWellData_,
+                this->isIORank()
+    };
+
+    PackUnPackGroupAndNetworkValues packUnpackGroupAndNetworkData {
+        localGroupAndNetworkData,
+                this->globalGroupAndNetworkData_,
+                this->isIORank()
+    };
+
+    PackUnPackBlockData packUnpackBlockData {
+        localBlockData,
+                this->globalBlockData_,
+                this->isIORank()
+    };
+
+    PackUnPackWBPData packUnpackWBPData {
+        localWBPData,
+                this->globalWBPData_,
+                this->isIORank()
+    };
+
+    PackUnPackAquiferData packUnpackAquiferData {
+        localAquiferData,
+                this->globalAquiferData_,
+                this->isIORank()
+    };
+
+    toIORankComm_.exchange(packUnpackCellData);
+    toIORankComm_.exchange(packUnpackWellData);
+    toIORankComm_.exchange(packUnpackGroupAndNetworkData);
+    toIORankComm_.exchange(packUnpackBlockData);
+    toIORankComm_.exchange(packUnpackWBPData);
+    toIORankComm_.exchange(packUnpackAquiferData);
+
+#ifndef NDEBUG
+    // make sure every process is on the same page
+    toIORankComm_.barrier();
+#endif
+}
+
+template <class Grid, class EquilGrid, class GridView>
+int CollectDataToIORank<Grid,EquilGrid,GridView>::
+localIdxToGlobalIdx(unsigned localIdx) const
+{
+    if (!isParallel())
+        return localIdx;
+
+    if (localIdxToGlobalIdx_.empty())
+        throw std::logic_error("index map is not created on this rank");
+
+    if (localIdx > localIdxToGlobalIdx_.size())
+        throw std::logic_error("local index is outside map range");
+
+    return localIdxToGlobalIdx_[localIdx];
+}
+
+template <class Grid, class EquilGrid, class GridView>
+bool CollectDataToIORank<Grid,EquilGrid,GridView>::
+isCartIdxOnThisRank(int cartIdx) const
+{
+    if (!isParallel())
+        return true;
+
+    assert(!needsReordering);
+    auto candidate = std::lower_bound(sortedCartesianIdx_.begin(), sortedCartesianIdx_.end(), cartIdx);
+    return (candidate != sortedCartesianIdx_.end() && *candidate == cartIdx);
+}
+
+
+
+template class CollectDataToIORank<Dune::CpGrid,
+                                   Dune::CpGrid,
+                                   Dune::GridView<Dune::DefaultLeafGridViewTraits<Dune::CpGrid>>>;
+template class CollectDataToIORank<Dune::PolyhedralGrid<3,3,double>,
+                                   Dune::PolyhedralGrid<3,3,double>,
+                                   Dune::GridView<Dune::PolyhedralGridViewTraits<3,3,double,Dune::PartitionIteratorType(4)>>>;
+
```

Comment:
I use the CollectDataToIORank in my adaptive gridding prototype simulator based on AluGrid.  Note the `needsReordering` I haven't read the code in details, but is it sufficient to add a AluGrid version here to make it compile and run with AluGrid?

Reply:
yes.

Reply:
>but in no way does it change the implementation
Good to know. It isn't always easy to tell from the `diff ` when it involves renaming.